pilot/pkg/model/push_context.go

```diff
@@ -1707,12 +1707,21 @@ func (ps *PushContext) BestEffortInferServiceMTLSMode(service *Service, port *Po
 		return MTLSUnknown
 	}
 
-	// First , check mTLS settings from beta policy (i.e PeerAuthentication) at namespace / mesh level.
+	// 1. Check service instances' tls mode, mainly used for headless service.
+	if service.Resolution != ClientSideLB {
+		instances := ps.InstancesByPort(service, port.Port, nil)
```

Comment:
I have some concerns about this call for performance. InstancesByPort is pretty slow if I remember right, may be ok but we should measure

Reply:
Yes, even with `if service.Resolution != ClientSideLB`, this can add calling (n*m) times. n is proxy number, m is the headless service number

Reply:
yikes.. this is pretty slow. should we look into caching the instances by port at each push epoch? just for the headless ones - dns, this pr , and listener code hit this call for each proxy.

Reply:
Yeah, this is also my thought. Created an issue to trace. https://github.com/istio/istio/issues/27117

Reply:
how is xDS of this code get triggered? Do we know if the pod come up or going down (with different label value changed), will trigger this code?

I remember something like `EDSUpdater` made some assumptions to not trigger things like CDS push when only pod is changing.

Reply:
Another way is to add some boolean flag to keep track whether all instances have mTLS label, and update this field when some endpoint is updated with different label value, and we know we need to trigger a push by then as well. We can only do that for the headless for performance sake.

```go
type Service {
  headlessAllIstioMTLSLabeled bool
  Endpoints []Endpoint
}
```

Just a thought for exploration.

Reply:
We donot need to handle labels at all, the pods are controlled by statefulset.  When the pods labels update, it means the pod is recreated, the endpoint event will trigger xds push.

Reply:
but i thought if it's only pod update, we won't trigger full xds push, only eds?

Reply:
It is not allowed to update the pod labels.

Reply:
Here we donot consider pods created directly, this applies to other features before.

Reply:
The calling to `ps.InstancesByPort` is being addressed, so this is not an issue